// styled-components-grid/dist/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _templateObject = _taggedTemplateLiteral(['\n  display: flex;\n  ', '\n  ', '\n  ', '\n  ', '\n'], ['\n  display: flex;\n  ', '\n  ', '\n  ', '\n  ', '\n']);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _styledComponentsBreakpoint = require('styled-components-breakpoint');

var _reactCreateComponentFromTagProp = require('react-create-component-from-tag-prop');

var _reactCreateComponentFromTagProp2 = _interopRequireDefault(_reactCreateComponentFromTagProp);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var GridComponent = (0, _reactCreateComponentFromTagProp2.default)({
  tag: 'div',
  prop: 'component',
  propsToOmit: ['wrap', 'reverse', 'horizontalAlign', 'verticalAlign']
});

var wrapMixin = function wrapMixin(_ref) {
  var wrap = _ref.wrap,
      theme = _ref.theme;
  return (0, _styledComponentsBreakpoint.map)(wrap, function () {
    var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    return 'flex-wrap: ' + (value && 'wrap' || 'nowrap') + ';';
  }, theme.breakpoints);
};

var flexDirectionMixin = function flexDirectionMixin(_ref2) {
  var reverse = _ref2.reverse,
      theme = _ref2.theme;
  return (0, _styledComponentsBreakpoint.map)(reverse, function (value) {
    return 'flex-direction: ' + (value && 'row-reverse' || 'row') + ';';
  }, theme.breakpoints);
};

var justifyContentMixin = function justifyContentMixin(_ref3) {
  var horizontalAlign = _ref3.horizontalAlign,
      reverse = _ref3.reverse,
      theme = _ref3.theme;
  return (0, _styledComponentsBreakpoint.map)(horizontalAlign, function (value) {
    var rule = '';
    switch (value) {

      default:
      case 'left':
        if (reverse) {
          rule = 'flex-end';
        } else {
          rule = 'flex-start';
        }
        break;

      case 'right':
        if (reverse) {
          rule = 'flex-start';
        } else {
          rule = 'flex-end';
        }
        break;

      case 'center':
        rule = 'center';
        break;

      case 'justify-center':
        rule = 'space-around';
        break;

      case 'justify':
        rule = 'space-between';
        break;

    }
    return 'justify-content: ' + rule + ';';
  }, theme.breakpoints);
};

var alignItemsMixin = function alignItemsMixin(_ref4) {
  var verticalAlign = _ref4.verticalAlign,
      theme = _ref4.theme;
  return (0, _styledComponentsBreakpoint.map)(verticalAlign, function (value) {
    var rule = '';
    switch (value) {

      case 'top':
        rule = 'flex-start';
        break;

      case 'bottom':
        rule = 'flex-end';
        break;

      case 'center':
        rule = 'center';
        break;

      default:
      case 'stretch':
        rule = 'stretch';
        break;

    }
    return 'align-items: ' + rule + ';';
  }, theme.breakpoints);
};

var Grid = (0, _styledComponents2.default)(GridComponent)(_templateObject, wrapMixin, flexDirectionMixin, justifyContentMixin, alignItemsMixin);

Grid.propTypes = {
  wrap: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.object]),
  reverse: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.object]),
  horizontalAlign: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['left', 'right', 'center', 'justify-center', 'justify']), _propTypes2.default.object]),
  verticalAlign: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['top', 'bototm', 'center', 'stretch']), _propTypes2.default.object])
};

Grid.defaultProps = {
  wrap: true,
  reverse: false,
  horizontalAlign: 'left',
  verticalAlign: 'stretch'
};

exports.default = Grid;
